app/models/concerns/labware_creators/donor_pooling_calculator.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+# This module contains calculations for donor pooling.
+module LabwareCreators::DonorPoolingCalculator
+  extend ActiveSupport::Concern
+
+  # Splits wells into groups by study and project. Wells are grouped
+  # based on the study and project of the first aliquot in each well. Returns
+  # an array of groups, where each group is an array of wells with the same
+  # study and project.
+  #
+  # @param group [Array<Well>] The group of wells to be split.
+  # @return [Array<Array<Well>>] An array of well groups.
+  def split_single_group_by_study_and_project(group)
+    group.group_by { |well| [well.aliquots.first.study.id, well.aliquots.first.project.id] }.values
+  end
+
+  # Splits groups ensuring unique donor_ids within each group. Iterates over
+  # each group, creating subgroups with wells from a unique donor. The first
+  # occurrences of unique donor_ids are grouped, then the second occurrences,
+  # and so on. This prevents combining samples with the same donor_id.
+  #
+  # @param groups [Array<Array<Well>>] Array of well groups to be split.
+  # @return [Array<Array<Well>>] Array of subgroups split by donor ID.
+  def split_groups_by_unique_donor_ids(groups)
+    groups.flat_map { |group| split_single_group_by_unique_donor_ids(group) }
+  end
+
+  # Splits a single group of wells by donor_ids. This method is used by the
+  # 'split_groups_by_unique_donor_ids' method. It iteratively segregates wells with
+  # the first encountered instance of each unique donor_id into a separate
+  # subgroup. This process continues until there are no wells left in the
+  # original group. The result is a collection of subgroups, each containing
+  # wells from distinct donors.
+  #
+  # @param group [Array<Well>] The group of wells to split.
+  # @return [Array<Array<Well>>] An array of subgroups, each containing wells
+  #   from different donors.
+  def split_single_group_by_unique_donor_ids(group)
+    group = group.dup
+    output = []
+    while group.any?
+      subgroup = []
+      unique_donor_ids(group).each do |donor_id|
+        index = group.index { |well| well.aliquots.first.sample.sample_metadata.donor_id == donor_id }
+        subgroup << group.delete_at(index) unless index.nil?
```

Comment:
Can `index` be `nil`? The `donor_id` should always be present in the group, right, because the list of donor_ids was retrieved from the group in the first place?

Reply:
You are right. It can't be nil as the unique_donor_ids is are calculated before each subgroup. Calculating before each subgroup seemed more performant if the donor ids are less repeated (more unique) as it reduces the number of iterations over the group. I am removing the redundant nil-check.  

I have been experimenting with another version where  unique_donor_ids are calculated only once. In that version of the algorithm we needed to check for nil. The nil-check is left-over from that version. It was causing more cycles over groups.